torch/nn/functional.py

```diff
@@ -5192,6 +5192,9 @@ def multi_head_attention_forward(
           :math:`S` is the source sequence length. If ``average_attn_weights=False``, returns attention weights per
           head of shape :math:`(num_heads, L, S)` when input is unbatched or :math:`(N, num_heads, L, S)`.
     """
+    # Ensure that `is_causal` is an actual bool and not a tensor containing a bool, which happens during tracing.
```

Comment:
Could this to be fixed on the ONNX tracing side instead of within this function?

`nn.MultiHeadAttention` expects `is_causal` to be a bool in the signature

Reply:
Right now we're raising a `RuntimeError` [here](https://github.com/pytorch/pytorch/blob/main/torch/jit/_trace.py#L118) if the arguments are not of type `torch.Tensor` during tracing. According to the [docs](https://pytorch.org/docs/stable/onnx.html#tracing-vs-scripting), it really has to be a `Tensor`. The C++ source for this underlying conversion to tensor is [here](https://github.com/pytorch/pytorch/blob/main/torch/csrc/jit/python/python_arg_flatten.cpp#L76), but changing that would probably be a larger effort.

Maybe we could have the native function `scaled_dot_product_attention()` accept a `tensor(bool)` instead.

Reply:
AFAIK TorchScript Tracing (which is leveraged by ONNX) cannot track non-Tensor types. @eellison might have an insight on how to work this around